forest/src/main.rs

```diff
@@ -6,23 +6,111 @@ mod daemon;
 mod logger;
 mod subcommand;
 
-use cli::{cli_error_and_die, Cli};
+use cli::{cli_error_and_die, Cli, DaemonConfig};
+
+use async_std::task;
+use daemonize_me::{Daemon, DaemonError, Group, User};
+use log::{error, info};
+use raw_sync::{events::*, Timeout};
+use shared_memory::{Shmem, ShmemConf};
 use structopt::StructOpt;
 
-#[async_std::main]
-async fn main() {
+use std::fs::File;
+use std::mem;
+use std::process;
+use std::sync::atomic::{AtomicPtr, Ordering};
+use std::time::Duration;
+
+static SHMEM_PTR: AtomicPtr<u8> = AtomicPtr::new(std::ptr::null_mut());
+
+const EVENT_TIMEOUT: Timeout = Timeout::Val(Duration::from_secs(4));
+
+fn create_event() -> Shmem {
+    let shmem = ShmemConf::new()
+        .size(mem::size_of::<Event>())
+        .create()
+        .expect("create must succeed");
+    SHMEM_PTR.store(shmem.as_ptr(), Ordering::Relaxed);
+    info!("Creating event in shared memory");
+    unsafe {
+        Event::new(shmem.as_ptr(), true).expect("new must succeed");
+    }
+    shmem
+}
+
+fn build_daemon<'a>(config: &DaemonConfig) -> Result<Daemon<'a>, DaemonError> {
+    let mut daemon = Daemon::new().umask(config.umask).work_dir(&config.work_dir);
+    if let Some(user) = &config.user {
+        daemon = daemon.user(User::try_from(user)?)
+    }
+    if let Some(group) = &config.group {
+        daemon = daemon.group(Group::try_from(group)?)
+    }
+    if let Some(path) = &config.stdout {
+        let file = File::create(path).expect("File creation {path} must succeed");
+        daemon = daemon.stdout(file)
+    }
+    if let Some(path) = &config.stderr {
+        let file = File::create(path).expect("File creation {path} must succeed");
+        daemon = daemon.stderr(file)
+    }
+    if let Some(path) = &config.pid_file {
+        daemon = daemon.pid_file(path, Some(false))
+    }
+
+    daemon = daemon.setup_post_fork_parent_hook(|_parent_pid, _child_pid| {
+        let (event, _) = unsafe {
+            Event::from_existing(SHMEM_PTR.load(Ordering::Relaxed))
+                .expect("from_existing must succeed")
+        };
+        if let Err(e) = event.wait(EVENT_TIMEOUT) {
+            error!("Event error: {e}");
+            process::exit(1);
+        }
+
+        info!("Exiting");
+
+        process::exit(0);
+    });
+
+    Ok(daemon)
+}
+
+fn main() {
     logger::setup_logger();
+
     // Capture Cli inputs
     let Cli { opts, cmd } = Cli::from_args();
 
     // Run forest as a daemon if no other subcommands are used. Otherwise, run the subcommand.
     match opts.to_config() {
         Ok(cfg) => match cmd {
-            Some(command) => subcommand::process(command, cfg).await,
-            None => daemon::start(cfg).await,
+            Some(command) => {
+                task::block_on(subcommand::process(command, cfg));
+            }
+            None => {
+                let _shmem = if opts.detach {
```

Comment:
Why do you return `shmem`?

Reply:
Otherwise the shared memory segment is dropped just before calling `::process` and forest woes.

Reply:
Ah sorry, that was for the `ShmemConf`. Investigating.

Reply:
Drop method is in `MapData` and will call `munmap`.

Reply:
Hmm, I think we have to be more careful about when `Shmem` is dropped. Maybe we need `lazy_static`. Then we can set the owner to be the parent.

Reply:
Right now it works mostly by accident. We should be explicit about dropping `Shmen` in the parent process, not in the child process. Doing this will mostly likely require a slight refactor.

Reply:
Can't we just create the value on the stack and let RAII take care of dropping the value?

Reply:
Not if we want to drop the reference from the parent. We can decide that it's fine to keep the reference around until the child dies, though.